File: artifacts/tulipindicators/decompose/src/psar.c
int ti_psar_start(const double *options)
{
  (void) options;
  return 1;
}


/*** DEPENDENCIES:
***/


// hint:  ['output_idx_ref is a mutable refrence to unsigned int', 'lng_ref is a mutable refrence to int', 'sar_ref is a mutable refrence to double', 'extreme_ref is a mutable refrence to double', 'accel_ref is a mutable refrence to double']
void helper_ti_psar_1(unsigned int * const output_idx_ref, int * const lng_ref, double * const sar_ref, double * const extreme_ref, double * const accel_ref, const double * const high, const double * const low, const double accel_step, const double accel_max, double * const output, int i)
{
  unsigned int output_idx = *output_idx_ref;
  int lng = *lng_ref;
  double sar = *sar_ref;
  double extreme = *extreme_ref;
  double accel = *accel_ref;
  sar = ((extreme - sar) * accel) + sar;
  if (lng)
  {
    if ((i >= 2) && (sar > low[i - 2]))
    {
      sar = low[i - 2];
    }
    if (sar > low[i - 1])
    {
      sar = low[i - 1];
    }
    if ((accel < accel_max) && (high[i] > extreme))
    {
      accel += accel_step;
      if (accel > accel_max)
      {
        accel = accel_max;
      }
    }
    if (high[i] > extreme)
    {
      extreme = high[i];
    }
  }
  else
  {
    if ((i >= 2) && (sar < high[i - 2]))
    {
      sar = high[i - 2];
    }
    if (sar < high[i - 1])
    {
      sar = high[i - 1];
    }
    if ((accel < accel_max) && (low[i] < extreme))
    {
      accel += accel_step;
      if (accel > accel_max)
      {
        accel = accel_max;
      }
    }
    if (low[i] < extreme)
    {
      extreme = low[i];
    }
  }
  if ((lng && (low[i] < sar)) || ((!lng) && (high[i] > sar)))
  {
    accel = accel_step;
    sar = extreme;
    lng = !lng;
    if (!lng)
    {
      extreme = low[i];
    }
    else
      extreme = high[i];
  }
  output[output_idx] = sar;
  output_idx += 1;
  *output_idx_ref = output_idx;
  *lng_ref = lng;
  *sar_ref = sar;
  *extreme_ref = extreme;
  *accel_ref = accel;
}


/*** DEPENDENCIES:
***/


int ti_psar(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  const double accel_step = options[0];
  const double accel_max = options[1];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (accel_step <= 0)
  {
    return 1;
  }
  if (accel_max <= accel_step)
  {
    return 1;
  }
  if (size < 2)
  {
    return 0;
  }
  int lng;
  if ((high[0] + low[0]) <= (high[1] + low[1]))
  {
    lng = 1;
  }
  else
    lng = 0;
  double sar;
  double extreme;
  if (lng)
  {
    extreme = high[0];
    sar = low[0];
  }
  else
  {
    extreme = low[0];
    sar = high[0];
  }
  double accel = accel_step;
  int i;
  for (i = 1; i < size; i += 1)
  {
    helper_ti_psar_1(&output_idx, &lng, &sar, &extreme, &accel, high, low, accel_step, accel_max, output, i);
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_psar_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_psar_start(const double *options)
{
  (void) options;
  return 1;
}


----------------------------
void helper_ti_psar_1(unsigned int * const output_idx_ref, int * const lng_ref, double * const sar_ref, double * const extreme_ref, double * const accel_ref, const double * const high, const double * const low, const double accel_step, const double accel_max, double * const output, int i)
{
  unsigned int output_idx = *output_idx_ref;
  int lng = *lng_ref;
  double sar = *sar_ref;
  double extreme = *extreme_ref;
  double accel = *accel_ref;
  sar = ((extreme - sar) * accel) + sar;
  if (lng)
  {
    if ((i >= 2) && (sar > low[i - 2]))
    {
      sar = low[i - 2];
    }
    if (sar > low[i - 1])
    {
      sar = low[i - 1];
    }
    if ((accel < accel_max) && (high[i] > extreme))
    {
      accel += accel_step;
      if (accel > accel_max)
      {
        accel = accel_max;
      }
    }
    if (high[i] > extreme)
    {
      extreme = high[i];
    }
  }
  else
  {
    if ((i >= 2) && (sar < high[i - 2]))
    {
      sar = high[i - 2];
    }
    if (sar < high[i - 1])
    {
      sar = high[i - 1];
    }
    if ((accel < accel_max) && (low[i] < extreme))
    {
      accel += accel_step;
      if (accel > accel_max)
      {
        accel = accel_max;
      }
    }
    if (low[i] < extreme)
    {
      extreme = low[i];
    }
  }
  if ((lng && (low[i] < sar)) || ((!lng) && (high[i] > sar)))
  {
    accel = accel_step;
    sar = extreme;
    lng = !lng;
    if (!lng)
    {
      extreme = low[i];
    }
    else
      extreme = high[i];
  }
  output[output_idx] = sar;
  output_idx += 1;
  *output_idx_ref = output_idx;
  *lng_ref = lng;
  *sar_ref = sar;
  *extreme_ref = extreme;
  *accel_ref = accel;
}


----------------------------
***/

File: artifacts/tulipindicators/translate/src/psar.rs
use crate::*;
use std::assert;

pub fn ti_psar_start(options: Option<&[f64]>) -> i32 {
    let _ = options; // Explicitly ignore the parameter to match C behavior
    1
}
pub fn helper_ti_psar_1(
    output_idx_ref: &mut u32,
    lng_ref: &mut i32,
    sar_ref: &mut f64,
    extreme_ref: &mut f64,
    accel_ref: &mut f64,
    high: &[f64],
    low: &[f64],
    accel_step: f64,
    accel_max: f64,
    output: &mut [f64],
    i: usize,
) {
    let mut output_idx = *output_idx_ref;
    let mut lng = *lng_ref;
    let mut sar = *sar_ref;
    let mut extreme = *extreme_ref;
    let mut accel = *accel_ref;

    sar = ((extreme - sar) * accel) + sar;

    if lng != 0 {
        if (i >= 2) && (sar > low[i - 2]) {
            sar = low[i - 2];
        }
        if sar > low[i - 1] {
            sar = low[i - 1];
        }
        if (accel < accel_max) && (high[i] > extreme) {
            accel += accel_step;
            if accel > accel_max {
                accel = accel_max;
            }
        }
        if high[i] > extreme {
            extreme = high[i];
        }
    } else {
        if (i >= 2) && (sar < high[i - 2]) {
            sar = high[i - 2];
        }
        if sar < high[i - 1] {
            sar = high[i - 1];
        }
        if (accel < accel_max) && (low[i] < extreme) {
            accel += accel_step;
            if accel > accel_max {
                accel = accel_max;
            }
        }
        if low[i] < extreme {
            extreme = low[i];
        }
    }

    if (lng != 0 && (low[i] < sar)) || ((lng == 0) && (high[i] > sar)) {
        accel = accel_step;
        sar = extreme;
        lng = !lng;
        if lng == 0 {
            extreme = low[i];
        } else {
            extreme = high[i];
        }
    }

    output[output_idx as usize] = sar;
    output_idx += 1;

    *output_idx_ref = output_idx;
    *lng_ref = lng;
    *sar_ref = sar;
    *extreme_ref = extreme;
    *accel_ref = accel;
}
pub fn ti_psar(
    size: usize,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    outputs: Option<&mut [&mut [f64]]>,
) -> i32 {
    // Check for None inputs (equivalent to NULL checks in C)
    if inputs.is_none() || options.is_none() || outputs.is_none() {
        return 1;
    }
    let inputs = inputs.unwrap();
    let options = options.unwrap();
    let outputs = outputs.unwrap();

    // Extract slices from inputs and outputs
    let high = inputs[0];
    let low = inputs[1];
    let output = &mut outputs[0];

    // Extract options
    let accel_step = options[0];
    let accel_max = options[1];

    // Early returns for invalid conditions (lines 9-20)
    if accel_step <= 0.0 {
        return 1;
    }
    if accel_max <= accel_step {
        return 1;
    }
    if size < 2 {
        return 0;
    }

    // Determine initial lng value (lines 21-27)
    let mut lng = if (high[0] + low[0]) <= (high[1] + low[1]) {
        1
    } else {
        0
    };

    // Initialize sar and extreme (lines 28-39)
    let (mut sar, mut extreme) = if lng != 0 {
        (low[0], high[0])
    } else {
        (high[0], low[0])
    };

    let mut accel = accel_step;
    let mut output_idx = 0;

    // Main processing loop (lines 42-45)
    for i in 1..size {
        helper_ti_psar_1(
            &mut output_idx,
            &mut lng,
            &mut sar,
            &mut extreme,
            &mut accel,
            high,
            low,
            accel_step,
            accel_max,
            output,
            i,
        );
    }

    // Assertion check (line 47)
    assert!((output[output_idx as usize] - outputs[0][0]) == (size as f64 - ti_psar_start(Some(options)) as f64));

    0
}
